source/h2_decoder.c

```diff
@@ -923,6 +979,146 @@ static int s_state_fn_frame_unknown(struct aws_h2_decoder *decoder, struct aws_b
+
+    bool is_pseudoheader = header_field->name.ptr[0] == ':';
+    if (is_pseudoheader) {
+        if (current_block->pseudoheaders_done) {
+            DECODER_LOG(ERROR, decoder, "Pseudo-headers must appear before regular fields.");
+            goto malformed;
+        }
+
+        enum pseudoheader_name pseudoheader_enum = s_header_to_pseudoheader_name(name_enum);
+        if (pseudoheader_enum == PSEUDOHEADER_UNKNOWN) {
+            DECODER_LOG(ERROR, decoder, "Unrecognized pseudo-header");
+            goto malformed;
+        }
+
+        /* Ensure request pseudo-headers vs response pseudoheaders were sent appropriately.
+         * This also ensures that request and response pseudoheaders aren't being mixed. */
+        bool expect_request_pseudoheader = decoder->is_server || current_block->is_push_promise;
+        bool is_request_pseudoheader = pseudoheader_enum != PSEUDOHEADER_STATUS;
+        if (expect_request_pseudoheader != is_request_pseudoheader) {
+            DECODER_LOG(ERROR, decoder, "Received wrong type of pseudo-header");
+            goto malformed;
+        }
+
+        /* Protect against duplicates. */
+        if (current_block->pseudoheader_received[pseudoheader_enum]) {
+            DECODER_LOG(ERROR, decoder, "Pseudoheader occurred multiple times");
+            goto malformed;
+        }
+        current_block->pseudoheader_received[pseudoheader_enum] = true;
+
+        /* NOTE: We validate pseudo-headers collectively, after they all arrive, in s_flush_pseudoheaders */
+
+    } else { /* Else regular header-field. */
+
+        /* Regular header-fields come after pseudo-headers, so make sure pseudo-headers are flushed */
+        if (!current_block->pseudoheaders_done) {
+            if (s_flush_pseudoheaders(decoder)) {
+                return AWS_OP_ERR;
+            }
+
+            /* might have realized that header-block is malformed during flush */
+            if (current_block->malformed) {
+                goto malformed;
+            }
+        }
+
+        /* Validate header name (not necessary if string already matched against a known enum) */
+        if (name_enum == AWS_HTTP_HEADER_UNKNOWN) {
+            if (!aws_strutil_is_lowercase_http_token(header_field->name)) {
+                DECODER_LOG(ERROR, decoder, "Header name contains invalid characters");
+                goto malformed;
+            }
+        }
+
+        /* #TODO Validate characters used in header_field->value */
+    }
+
+    /* #TODO should we hold back pseudo-headers until s_flush_pseudoheaders() is called?
+     * - PRO: will never deliver pseudo-header that we later determine is invalid.
+     * - CON: buffering */
+
+    /* Deliver header-field via callback */
+    if (current_block->is_push_promise) {
+        DECODER_CALL_VTABLE_STREAM_ARGS(decoder, on_push_promise_i, header_field, name_enum);
+    } else {
+        DECODER_CALL_VTABLE_STREAM_ARGS(decoder, on_headers_i, header_field, name_enum);
```

Comment:
are we doing a callback per header?

Reply:
The public HTTP API is capable of delivering them in batches.

But this internal API is delivering them one at a time.
Though I **just** at Justin's suggestion changed it so we buffer up the pseudo-headers, so it would be theoretically more efficient to deliver all in a single callback.

Honestly everyone's life would be easier if we buffered all headers deep down here in the decoder and always delivered everything all at once. I know that's how we made the callbacks work in (some? most? all?) of the CRT language bindings. The only reason I don't 100% push for that change here in C is that it is less efficient for languages that will need to make a complete copy of the strings in order to pass them to the user

ramble ramble